Sources/JavaScriptEventLoop/JSAsyncFunction.swift

```diff
@@ -0,0 +1,68 @@
+import JavaScriptKit
+
+/// A `JSFunction` wrapper that enables throwing function calls.
+/// Exceptions produced by JavaScript functions will be thrown as `JSValue`.
+public class JSAsyncFunction {
+    private let base: JSFunction
+    public init(_ base: JSFunction) {
+        self.base = base
+    }
+
+    /// Call this function with given `arguments` and binding given `this` as context.
+    /// - Parameters:
+    ///   - this: The value to be passed as the `this` parameter to this function.
+    ///   - arguments: Arguments to be passed to this function.
+    /// - Returns: The result of this call.
+    @discardableResult
+    public func callAsFunction(this: JSObject? = nil, arguments: [ConvertibleToJSValue]) async throws -> JSValue {
+        let result = base.callAsFunction(this: this, arguments: arguments)
+        return await try JSPromise<JSValue, JSValue>(result.object!)!.await()
+    }
+
+    /// A variadic arguments version of `callAsFunction`.
+    @discardableResult
+    public func callAsFunction(this: JSObject? = nil, _ arguments: ConvertibleToJSValue...) async throws -> JSValue {
+        await try callAsFunction(this: this, arguments: arguments)
+    }
+}
+
+public extension JSFunction {
+    var `async`: JSAsyncFunction {
+        JSAsyncFunction(self)
+    }
+}
+
+/// A `JSObject` wrapper that enables throwing method calls capturing `this`.
+/// Exceptions produced by JavaScript functions will be thrown as `JSValue`.
+@dynamicMemberLookup
+public class JSAsyncingObject {
+    private let base: JSObject
+    public init(_ base: JSObject) {
+        self.base = base
+    }
+
+    /// Returns the `name` member method binding this object as `this` context.
+    /// - Parameter name: The name of this object's member to access.
+    /// - Returns: The `name` member method binding this object as `this` context.
+    @_disfavoredOverload
+    public subscript(_ name: String) -> ((ConvertibleToJSValue...) async throws -> JSValue)? {
```

Comment:
Would it be possible to implement this as an overload in the regular `JSObject`? Not sure how overloads work exactly in this case but it would lead to a much nicer DX if possible.

Reply:
Unfortunately, `async` and non-`async` function overloads are ambiguous 😢